quic/s2n-quic/src/provider/endpoint_limits.rs

```diff
@@ -29,6 +33,125 @@ impl<T: 'static + Limiter> Provider for T {
     }
 }
 
+const THROTTLED_PORT_LIMIT: usize = 10;
+const THROTTLE_FREQUENCY_MS: u64 = 1000;
+
+#[derive(Default, Debug, Clone, Copy)]
+struct BasicRateLimiter {
+    last_throttle_reset: Option<Timestamp>,
+    count: usize,
+}
+
+impl BasicRateLimiter {
+    /// Throttles a connection based on a limit.
+    /// Returns True if the `should_throttle` invoke count is greater than `limit` and the
+    /// connection has not been throttled in the last `throttle_frequency` duration.
+    ///
+    /// If the throttle timer expires the count is reset and `should_throttle` returns False.
+    ///
+    /// Returns False if the `should_throttle` invoke count is less than `limit`.
+    fn should_throttle(
+        &mut self,
+        limit: usize,
+        throttle_frequency: Duration,
+        connection_attempt: &ConnectionAttempt,
+    ) -> bool {
+        self.count += 1;
+        let time_stamp = connection_attempt.time_stamp;
+
+        if self.count > limit {
+            match self.last_throttle_reset {
+                // If the throttle timer is still within the throttle_frequency
+                // then throttle the connection.
+                Some(last_throttle_reset)
+                    if time_stamp.saturating_duration_since(last_throttle_reset)
+                        < throttle_frequency =>
+                {
+                    return true;
+                }
+                // If the throttle timer is greater than the throttle_frequency
+                // then reset the throttle count and the throttle timer.
+                // Let the connection through.
+                _ => {
+                    self.count = 0;
+                    self.last_throttle_reset = Some(time_stamp);
+                    return false;
+                }
+            };
+        }
+
+        // If this is the first time calling instantiate the throttle timer.
+        if self.last_throttle_reset.is_none() {
+            self.last_throttle_reset = Some(time_stamp);
+        }
+
+        false
+    }
+}
+
+#[cfg(any(test, feature = "testing"))]
```

Comment:
I grabbed "testing" to align with the mock clock macro. Unfortunately I am cargo culting because I don't know what the use case is for this feature outside of making the code only available in testing.

Reply:
we only use the `feature = "testing"` when another test might want to depend on it some functionality inside. In this case we're not exporting anything so you only need `cfg(test)`

Reply:
Thanks!

Reply:
`git stash` :(. Thanks for catching this (again)